### PR TITLE
Scramble banks from the end of the ROM

### LINKED ISSUE #1036: Update scramble to specially handle >8-bit banks
(Opened since https://github.com/gbdev/rgbds/pull/1014 was accidentally deleted)

the -S flag works great for a single byte, but for any roms over 2MiB it should also consider the upper byte when scrambling

### PR DESCRIPTION
**Pull request by Rangi42**
This is more likely to test edge cases, such as having content in banks with their highest bit set.

Fixes #1036.

#1149 proposes allocating sections from the *end* of their individual banks. This is a similar approach, allocating section banks from the end of their overall ROM space.

### CODE REVIEW

File: test/link/test.sh [hunk 1]
```diff
@@ -106,6 +106,13 @@ done
 
 # These tests do their own thing
 
+i="scramble-romx.asm"
+startTest
+"$RGBASM" -o "$otemp" scramble-romx/a.asm
+rgblinkQuiet -o "$gbtemp" -S romx=3 "$otemp" >"$outtemp" 2>&1
+tryDiff scramble-romx/out.err "$outtemp"
+(( rc = rc || $? ))
```
Comment (ISSOtm) [inline test/link/test.sh:114]: This should also check that the output binary is 4 banks big.

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: test/link/test.sh (hunk 1 above)
Comment (Rangi42) [reply test/link/test.sh]: It is not! :P

**Review by ISSOtm — DISMISSED** [review]
Looks good to me!

**Comment by Rangi42** [comment]
Okay, now trying the *next* bank is updated too, not just assigning the *first* bank. Say you pass `-S romx=3`: then it will try the ROMX banks in order 3, 2, 1, 4, 5, 6, 7, .... Or `-S sram=4`: they'll go 4, 3, 2, 1, 0, 5, 6, 7, ....

File: test/link/test.sh (hunk 1 above)
Comment (Rangi42) [reply test/link/test.sh]: It is now. :)

File: src/link/assign.cpp [hunk 2]
```diff
@@ -179,20 +178,43 @@ static struct FreeSpace *getPlacement(struct Section const *section,
...
+				return NULL;
+		} else if (scrambleSRAM && section->type == SECTTYPE_SRAM && location->bank <= scrambleSRAM) {
+			if (location->bank > sectionTypeInfo[section->type].firstBank)
+				location->bank--;
+			else if (scrambleSRAM < sectionTypeInfo[section->type].lastBank)
+				location->bank = scrambleSRAM + 1;
+			else
+				return NULL;
+		} else if (location->bank < sectionTypeInfo[section->type].lastBank) {
+			location->bank++;
```
Comment (ISSOtm) [inline src/link/assign.cpp:214]: I hate that all of this logic is largely duplicated, and also feels very "action at a distance" (the logic that requires *this* to switch directions is far away). But I can swallow that pill considering that this logic isn't in our backwards compat promise (so we can implement other strategies) and we're going to rewrite this soon anyway.

File: test/link/test.sh [hunk 3]
```diff
@@ -146,6 +146,16 @@ tryDiff overlay/out.err "$outtemp"
...
 (( rc = rc || $? ))
 
+i="scramble-romx.asm"
+startTest
+"$RGBASM" -o "$otemp" scramble-romx/a.asm
+rgblinkQuiet -o "$gbtemp" -S romx=3 "$otemp" >"$outtemp" 2>&1
+tryDiff scramble-romx/out.err "$outtemp"
+(( rc = rc || $? ))
+# This test does not compare its exact output with 'tryCmpRom' because no scrambling order is guaranteed
+test $(wc -c <"$gbtemp") = 65536
```
Comment (ISSOtm) [inline test/link/test.sh:156]:
```suggestion
test "$(wc -c <"$gbtemp")" = 65536 # Check for exactly 3 ROMX banks.
```

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: src/link/assign.cpp (hunk 2 above)
Comment (Rangi42) [reply src/link/assign.cpp:214]: I don't like it either, but didn't see a way to abstract out the repetition that wouldn't cost more in complexity than it saves.

Comment (ISSOtm) [reply src/link/assign.cpp:214]: Not in C, anyway. (Macros could be involved, but let's not.)

**Review by ISSOtm — APPROVED** [review]
Subject to testing by @eievui5 before release, preferably?

**Comment by Rangi42** [comment]
The macOS tests give this error:
```
./test.sh: line 14: rm -f ${otemp@Q} ${gbtemp@Q} ${gbtemp2@Q} ${outtemp@Q}: bad substitution
```

It's complaining about this line:
```bash
trap "rm -f ${otemp@Q} ${gbtemp@Q} ${gbtemp2@Q} ${outtemp@Q}" EXIT
```

I haven't changed the definitions of those temp files, so not sure what the actual problem is.

**Comment by Rangi42** [comment]
> Subject to testing by @eievui5 before release, preferably?

Tested and approved!
